x-pack/plugins/infra/public/containers/logs/log_position/log_position_state.ts

```diff
@@ -79,26 +114,94 @@ export const useLogPositionState: () => LogPositionStateParams & LogPositionCall
     [startKey, endKey]
   );
 
+  // Allow setting `startDate` and `endDate` separately, or together
+  const updateDateRange = useCallback(
+    (newDateRange: Partial<DateRange>) => {
+      // Prevent unnecessary re-renders
+      if (!('startDateExpression' in newDateRange) && !('endDateExpression' in newDateRange)) {
+        return;
+      }
+
+      const nextStartDateExpression =
+        newDateRange.startDateExpression || dateRange.startDateExpression;
+      const nextEndDateExpression = newDateRange.endDateExpression || dateRange.endDateExpression;
+
+      if (!isValidDatemath(nextStartDateExpression) || !isValidDatemath(nextEndDateExpression)) {
+        return;
+      }
+
+      // Dates are valid, so the function cannot return `null`
+      const nextStartTimestamp = datemathToEpochMillis(nextStartDateExpression)!;
+      const nextEndTimestamp = datemathToEpochMillis(nextEndDateExpression, 'up')!;
+
+      // Reset the target position if it doesn't fall within the new range.
+      if (
+        targetPosition &&
+        (nextStartTimestamp > targetPosition.time || nextEndTimestamp < targetPosition.time)
+      ) {
+        jumpToTargetPosition(null);
+      }
+
+      setDateRange(previousDateRange => {
+        return {
+          startDateExpression:
+            newDateRange.startDateExpression || previousDateRange.startDateExpression,
+          endDateExpression: newDateRange.endDateExpression || previousDateRange.endDateExpression,
```

Comment:
Further up we're accessing the `dateRange.startDateExpression` and `dateRange.endDateExpression` directly from the outer scope while we use the `previousDateRange` here. I too prefer the latter style, but I wonder if we might run into inconsistencies? :thinking:

Reply:
> Further up we're accessing the dateRange.startDateExpression and dateRange.endDateExpression directly from the outer scope

I'm not sure I understand what you mean here 😳

Reply:
This uses the function form of `setState`, which gets passed the previous state, which is then used to partial update the date expressions. In the conditions leading up to it we use `dateRange` from the scope of the containing function, though. Since `setState` is asynchronous, though, `dateRange.startDateExpression` in the conditions before the `setState` and `previousDateRange.startDateExpression` within the updater function may be different. So the decision to assign a value might have been made based on a different value than what will ultimately be assigned.

I doesn't seem to be a problem right now, but future implementation changes in the React scheduler might lead to a subtle data race here, which will be hard to debug.

Reply:
Ok I follow now.

I think i can use the [`useSetState()`](https://streamich.github.io/react-use/?path=/story/state-usesetstate--docs) hook from react-use and get rid of the callback. The purpose of the callback is the same